p2p/subscriber_metrics.go

```diff
@@ -0,0 +1,130 @@
+package p2p
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+const (
+	statusKey    = "status"
+	statusAccept = "accept"
+	statusIgnore = "ignore"
+	statusReject = "reject"
+)
+
+type subscriberMetrics struct {
+	messageNumInst  metric.Int64Counter
+	messageSizeInst metric.Int64Histogram
+
+	messageTimeLastMu sync.Mutex
+	messageTimeLast   time.Time
+	messageTimeInst   metric.Float64Histogram
+
+	subscriptionNum     atomic.Int64
+	subscriptionNumInst metric.Int64ObservableGauge
+	subscriptionNumReg  metric.Registration
+}
+
+func newSubscriberMetrics() (m *subscriberMetrics, err error) {
+	m = new(subscriberMetrics)
+	m.messageNumInst, err = meter.Int64Counter(
+		"hdr_p2p_sub_msg_num",
+		metric.WithDescription("header message count"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.messageSizeInst, err = meter.Int64Histogram(
+		"hdr_p2p_sub_msg_size",
+		metric.WithDescription("valid header message size"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.messageTimeInst, err = meter.Float64Histogram(
+		"hdr_p2p_sub_msg_time",
+		metric.WithDescription("valid header message propagation time"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.subscriptionNumInst, err = meter.Int64ObservableGauge(
+		"hdr_p2p_sub_num",
+		metric.WithDescription("number of active header message subscriptions"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	m.subscriptionNumReg, err = meter.RegisterCallback(m.subscriptionCallback, m.subscriptionNumInst)
+	if err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+func (m *subscriberMetrics) accept(ctx context.Context, size int) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.messageNumInst.Add(ctx, 1, metric.WithAttributes(
+			attribute.String(statusKey, statusAccept),
+		))
+		m.messageSizeInst.Record(ctx, int64(size))
+
+		now := time.Now()
+		m.messageTimeLastMu.Lock()
+		if !m.messageTimeLast.IsZero() {
+			m.messageTimeInst.Record(ctx, now.Sub(m.messageTimeLast).Seconds())
+		}
+		m.messageTimeLast = now
+		m.messageTimeLastMu.Unlock()
+	})
+}
+
+func (m *subscriberMetrics) ignore(ctx context.Context) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.messageNumInst.Add(ctx, 1, metric.WithAttributes(
+			attribute.String(statusKey, statusIgnore),
+		))
+	})
+}
+
+func (m *subscriberMetrics) reject(ctx context.Context) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.messageNumInst.Add(ctx, 1, metric.WithAttributes(
+			attribute.String(statusKey, statusReject),
+		))
+	})
+}
```

Comment:
Alternative / Optional:
Could be single function `observeStatus(ctx, status)`

Reply:
I would do that if we had the same code path for accepts, but as long as accept has to be its own method, keeping standalone ignore and reject is nicer for consistency